### DISCUSSION TITLE
TODO List

### DISCUSSION THREAD

**Discussion by Jaimi5** [LoRaMesher/LoRaMesher#35] [Announcements]
### Overview 
In this page, we will comment what should be done to accomplish a fully functional library.

Feel free to comment for new ideas!

### Features
- [x] Test timeout for routing tables 
- [x] Improve duty cycle, do it with the number of bytes sent
- [x] Add ulTaskNotifyTake at the sending task. If there are no send packets, just wait for someone to send one and not execute anything of the function.
- [x] Pause LoRaMesher
- [x] Add timeout for the large and reliable payloads
     - [x] Consider how many hops to reach a node
     - [x] Consider duty cycle
     - [x] Consider traffic delays
     - [x] Test timeouts

- [ ] Refactor the code
- [ ] Add multiple entries in the routing table
- [ ] Add and share more metrics to the routing table

- [ ] FreeRTOS
     - [ ] Use Queues from FreeRTOS
     - [x] Use memory allocation from FreeRTOS

- [ ] When sending a message, assign an id and return to user. When send notice the user that has been send.
- [ ] Calculate energy consumption of the library
- [ ] Find best configuration values for the library
- [ ] Implement Round Robin inside send queue priorities. Avoid sending all the time the packets with lowest priorities.
- [ ] Add end-to-end encryption
- [ ] AODV instead of DSDV

### Bugs
- [ ] Add millis() check, max time 49 days now
- [x] Send packets delay. Try to fix packet collisions.

### Future ideas
- [ ] A real application example using LoRaMesher
- [ ] Add address filtering in LoRa packets. It will improve the battery life.
- [ ] Add Traceroute

### Some questions
- [ ] Allow users to send large payloads without payload (Like a ping?). Could be used to find the round-trip delay time.
- [ ] Could it be possible that users specify different configuration values? Timeouts could be different then and other things could broke.

**Comment by PureTek-Innovations** [comment]
Hi
This looks very interesting, thank you for publishing it.
I have a real-world use for it.
I'm very happy to test, but my coding skills are mostly limited to combining and troubleshooting other people's code.
I'll send you an email.
Thanks again
